planner/core/testdata/rule_constant_propagation_suite_out.json

```diff
@@ -0,0 +1,243 @@
+          "HashJoin_10 4166.67 root  right outer join, equal:[eq(test.t.id, test.s.id)]",
+          "├─TableReader_17(Build) 3333.33 root  data:Selection_16",
+          "│ └─Selection_16 3333.33 cop[tikv]  gt(test.s.id, 1)",
+          "│   └─TableFullScan_15 10000.00 cop[tikv] table:s keep order:false, stats:pseudo",
+          "└─TableReader_14(Probe) 3333.33 root  data:Selection_13",
+          "  └─Selection_13 3333.33 cop[tikv]  gt(test.t.id, 1), not(isnull(test.t.id))",
+          "    └─TableFullScan_12 10000.00 cop[tikv] table:t keep order:false, stats:pseudo"
+        ]
+      },
+      {
+        "SQL": "explain select * from (select * from t where t.id>1) tmp  right join s on tmp.id=s.id; -- can't, right join",
+        "Output": [
+          "HashJoin_10 10000.00 root  right outer join, equal:[eq(test.t.id, test.s.id)]",
+          "├─TableReader_13(Build) 3333.33 root  data:Selection_12",
+          "│ └─Selection_12 3333.33 cop[tikv]  gt(test.t.id, 1), not(isnull(test.t.id))",
+          "│   └─TableFullScan_11 10000.00 cop[tikv] table:t keep order:false, stats:pseudo",
+          "└─TableReader_15(Probe) 10000.00 root  data:TableFullScan_14",
+          "  └─TableFullScan_14 10000.00 cop[tikv] table:s keep order:false, stats:pseudo"
+        ]
+      }
+    ]
+  },
+  {
+    "Name": "TestSelectionThroughPlanNode",
+    "Cases": [
+      {
+        "SQL": "explain select * from t, (select id as id1 from s where s.id>1) tmp where t.id=tmp.id1; -- constant propagation can through the projection node",
+        "Output": [
+          "HashJoin_11 4166.67 root  inner join, equal:[eq(test.t.id, test.s.id)]",
+          "├─TableReader_18(Build) 3333.33 root  data:Selection_17",
+          "│ └─Selection_17 3333.33 cop[tikv]  gt(test.s.id, 1), not(isnull(test.s.id))",
+          "│   └─TableFullScan_16 10000.00 cop[tikv] table:s keep order:false, stats:pseudo",
+          "└─TableReader_15(Probe) 3333.33 root  data:Selection_14",
+          "  └─Selection_14 3333.33 cop[tikv]  gt(test.t.id, 1), not(isnull(test.t.id))",
+          "    └─TableFullScan_13 10000.00 cop[tikv] table:t keep order:false, stats:pseudo"
+        ]
+      },
+      {
+        "SQL": "explain select * from t, (select id, count(name) from s where s.id>1 group by id) tmp where t.id=tmp.id; -- can't, constant propagation can't through the aggregation node",
+        "Output": [
+          "Projection_11 3333.33 root  test.t.id, test.t.name, test.s.id, Column#7",
+          "└─Projection_12 3333.33 root  test.t.id, test.t.name, Column#7, test.s.id",
+          "  └─HashJoin_14 3333.33 root  inner join, equal:[eq(test.s.id, test.t.id)]",
+          "    ├─HashAgg_20(Build) 2666.67 root  group by:test.s.id, funcs:count(Column#8)->Column#7, funcs:firstrow(test.s.id)->test.s.id",
+          "    │ └─TableReader_21 2666.67 root  data:HashAgg_15",
+          "    │   └─HashAgg_15 2666.67 cop[tikv]  group by:test.s.id, funcs:count(test.s.name)->Column#8",
```

Comment:
group by id should be work?  the condition can be throw out from group by items

Reply:
Yes. But this time I don't support it. Because more business scenarios do not need to through agg. However, I have already designed the interface. If we need it later, we can just implement an interface directly.